Initial version of GKE Autopilot documentation.

naemono opened the pull request:
related #6713 

This change adds a page to our documentation that discusses running ECK, Elasticsearch and Kibana on GKE Autopilot clusters. Any of the other supported stack (Beat/Agent/etc) aren't discussed at this time as we have not verified their functionality as of yet.

This also adjusts the virtual memory documentation to note that there is an additional option for setting `max_map_count`.

barkbay commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,145 @@
+----
+
+[id="{p}-autopilot-deploy-kibana"]
+== Deploy a Kibana instance
```
nit: I think that as a user I would wonder why Kibana is mentioned, and not the other Stack components.

pebrc replied:
Why are we basically repeating the quickstart example her if there is no substantial difference for Autopilot?

barkbay replied:
👍 , I think I would remove that section.

barkbay commented on docs/orchestrating-elastic-stack-applications/elasticsearch/virtual-memory.asciidoc:
```diff
@@ -11,7 +11,7 @@ endif::[]
 Usually, default values for virtual address space on Linux distributions are too low for Elasticsearch to work properly, which may result in out-of-memory exceptions. This is why link:k8s-quickstart.html[the quickstart example] disables `mmap` through the `node.store.allow_mmap: false` setting. For production workloads, it is strongly recommended to increase the kernel setting `vm.max_map_count`  ...
 
-The kernel setting `vm.max_map_count=262144` can be set on the host either directly or by a dedicated init container, which must be privileged. To add an init container that changes the host kernel setting before your Elasticsearch pod starts, you can use the following example Elasticsearch spec:
+The kernel setting `vm.max_map_count=262144` can be set on the host directly, by a dedicated init container which must be privileged, or a dedicated Daemonset as seen in the <<{p}-autopilot>> section. To add an init container that changes the host kernel setting before your Elasticsearch pod starts, you can use the following example Elasticsearch spec:
```
As mentioned in my previous comment, I think I would do this the other way around because it can help on other types of K8S clusters:

## Virtual memory

By default, Elasticsearch uses memory mapping (mmap) to efficiently access indices.....

### Using an init container

The kernel setting vm.max_map_count=262144 can be set on the host directly, by a dedicated init container....

### Using a DaemonSet

The kernel setting can also be set on all the cluster nodes using a `DaemonSet` .... pros/cons ... This is the recommended approach on Google Autopilot<<{p}-autopilot>>.

naemono commented:
@barkbay I believe I've addressed your concerns in your previous review. I also added a section detailing the installation of Elastic Agent, but not the other stack components. Let me know if we should document all of them here.

barkbay commented on docs/orchestrating-elastic-stack-applications/elasticsearch/virtual-memory.asciidoc:
```diff
@@ -37,6 +41,69 @@ EOF
+EOF
+----
+
 For more information, check the Elasticsearch documentation on
```
Should we move this section before `Using an Init Container to set virtual memory` ?

barkbay commented on docs/orchestrating-elastic-stack-applications/elasticsearch/virtual-memory.asciidoc:
```diff
@@ -37,6 +41,69 @@ EOF
+          image: docker.io/bash:5.2.15
+          command: ['sleep', 'infinity']
+---
+apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
```
nit: I think I would split this manifest into 2 parts to make it more obvious what are their purposes.

naemono commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,281 @@
+                # cpu: 200m # Keep this commented. We dont set the cpu limit to avoid scheduling problems of agent in autopilot scenarios
+                memory: 700Mi
+                ephemeral-storage: "500Mi"
+        serviceAccountName: elastic-agent
```
This is why I hesitated to mention agent in these installation docs, as the steps to install/run it are rather large for this documentation; see https://github.com/elastic/cloud-on-k8s/pull/6719/files#diff-4859960fe8ea6730af189e41297306b45dc4a3ce0ef12078ad74c5ce6626b975

pebrc replied:
Plus one to removing the example and maybe link to a recipe. Even if it means waiting for that other PR. I would focus this documenation chapter on the material differences that a user has to be aware of to successfully deploy on Autopilot.

naemono replied:
I've simply added the recipe to this PR, as #6719 is not required to run Agent within Autopilot.

naemono commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,281 @@
+== Before you begin
+
+. Ensure you are running a supported version of Kubernetes within Autopilot.
++
```
I think it was missing "float", but I've removed it. Strange it wasn't showing up my preview pane, but was showing up like this when actually generating the docs.

barkbay commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,259 @@
+[id="{p}-autopilot-deploy-elasticsearch"]
+== Deploy an Elasticsearch instance
+
+Create an Elasticsearch cluster. If you are using the `Daemonset` described in the <<{p}-virtual-memory>> section to set `max_map_count` ensure that the `initContainer` below is also used to ensure the setting prior to starting Elasticsearch.
```
```suggestion
Create an Elasticsearch cluster. If you are using the `Daemonset` described in the <<{p}-virtual-memory>> section to set `max_map_count` you can add the `initContainer` below to ensure the setting is set prior to starting Elasticsearch.
```

barkbay commented on docs/orchestrating-elastic-stack-applications/elasticsearch/virtual-memory.asciidoc:
```diff
@@ -37,8 +44,74 @@ EOF
+          command: ['sleep', 'infinity']
+----
+
+== Deploy an Elasticsearch instance
```
```suggestion
```

pebrc commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,259 @@
+    podTemplate:
+      spec:
+        # This init container ensures that the `max_map_count` setting has been applied before starting Elasticsearch.
+        # This is not required, but is encouraged when using the previous Daemonset to set max_map_count.
```
```suggestion
        # This is not required, but is encouraged when using the previously mentioned Daemonset to set max_map_count.
```
?

pebrc commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,259 @@
+
+. Refer to <<{p}-installing-eck>> for more information on installation options.
+
+. Create a namespace to hold the Elastic resources ({eck_resources_list}):
```
Why are we including instructions to install the operator here that are identical to our regular instructions? What is the purpose of the instruction to create a dedicated namespace? is this an Autopilot requirement (I assume not)

barkbay replied:
> I'm following our example laid out in openshift.asciidoc where we also discuss installing the operator.

I think OpenShift is bit "special" in the sense that OpenShift users always create a "project" using `oc new-project` before creating other resources. That's why I added the `oc new-project elastic` in the case of OpenShift.

naemono replied:
Removed all references to creating a namespace and defaulted to the default namespace.

pebrc reviewed (commented):
I would focus on the essentials here. Also if I look at this documentation in its rendered form then most important part (the virtual memory bit) can be easily overlooked as it does not get its own sub-chapter 
<img width="1453" alt="Screenshot 2023-05-15 at 15 09 40" src="https://github.com/elastic/cloud-on-k8s/assets/697790/76b599c9-2a22-4edd-80d4-46824264701a">

If I understand correctly what is necessary to run Elasticsearch on autopilot then it's:
1. set virtual memory with the white-listed DaemonSet
2. you are done (but you could set this init container to improve the ES startup experience)

The extra init container to wait for it to be set is a nice to have and can be mentioned but should be qualified as such. 

3. Agent (but also Beats?)
Here I would describe the necessary changes (no hostpath volumes, but one can use ephemeral volumes or ephemeral storage requests instead, and you mentioned something about CPU limits) and link to an example recipe in the recipe folder

Let me know if I got something wrong.

naemono commented:
> 3. Agent (but also Beats?)
>     Here I would describe the necessary changes (no hostpath volumes, but one can use ephemeral volumes or ephemeral storage requests instead, and you mentioned something about CPU limits) and link to an example recipe in the recipe folder

I'm working through adding these additional bits to `gke-autopilot.asciidoc`....

I'll also work through testing some of the beats examples and getting recipes for them...

naemono commented:
@pebrc @barkbay I've tested multiple Agent configurations in Autopilot, and a Metricsbeat configuration and added them as example manifests. I've also updated the GKE Autopilot documentation for Agent/Beats noting the volume limitations, and linking to the recipes.

barkbay commented on config/recipes/autopilot/fleet-kubernetes-integration.yaml:
```diff
@@ -0,0 +1,263 @@
+      - logs
+      - metrics
+      unenroll_timeout: 900
+      is_default_fleet_server: true
```
```suggestion
```

barkbay commented on config/recipes/autopilot/fleet-kubernetes-integration.yaml:
```diff
@@ -0,0 +1,263 @@
+      - logs
+      - metrics
+      unenroll_timeout: 900
+      is_default: true
```
```suggestion
```

naemono commented:
> We should update `config/recipes/autopilot/fleet-kubernetes-integration.yaml` to include an explicit policy using `policyID`, see #6678

Thanks for noting this. I had missed this change.

barkbay commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,71 @@
+[id="{p}-autopilot-deploy-the-operator"]
+== Installing the ECK Operator
+
+. Refer to <<{p}-installing-eck>> for more information on installation options.
```
```suggestion
Refer to <<{p}-installing-eck>> for more information on installation options.
```

barkbay commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,71 @@
+This page shows how to run ECK on GKE Autopilot.
+
+1. It is recommended that each Kubernetes host's virtual memory kernel settings be modified. Refer to <<{p}-virtual-memory>>.
+2. It is recommended that Elasticsearch have an `initContainer` that waits for virtual memory settings to be in place. Refer to <<{p}-autopilot-deploy-elasticsearch>>.
```
```suggestion
2. It is recommended that Elasticsearch Pods have an `initContainer` that waits for virtual memory settings to be in place. Refer to <<{p}-autopilot-deploy-elasticsearch>>.
```

barkbay commented on docs/advanced-topics/gke-autopilot.asciidoc:
```diff
@@ -0,0 +1,71 @@
+[id="{p}-autopilot-deploy-agent-beats"]
+== Deploy a standalone Elastic Agent and/or Beats
+
+When running Elastic Agent and Beats within GKE Autopilot there are storage constraints to be considered.
+
+1. No `HostPath` volumes are allowed, which the ECK operator defaults to when unset for both `Deployments` and `Daemonsets`. Instead use link:https://kubernetes.io/docs/concepts/storage/ephemeral-volumes[Kubernetes ephemeral volumes].
```
```suggestion
When running Elastic Agent and Beats within GKE Autopilot there are storage constraints to be considered. No `HostPath` volumes are allowed, which the ECK operator defaults to when unset for both `Deployments` and `Daemonsets`. Instead use link:https://kubernetes.io/docs/concepts/storage/ephemeral-volumes[Kubernetes ephemeral volumes].
```

barkbay reviewed (approved):
LGTM
